File: leetcode/algorithms/28_find_the_index_of_the_first_occurrence_in_a_string/main.py
class FindTheIndexOfTheFirstOccurrenceInAString:
    """
    # Brute Force, Sliding Window
    #
    # Complexities:
    #   N - Length of `haystack`
    #   M - Length of `needle`
    #   - Time Complexity: O(N * M)
    #   - Space Complexity: O(1)
    """
    def strStr(self, haystack: str, needle: str) -> int:
        haystack_len: int = len(haystack)
        needle_len: int = len(needle)

        if needle_len > haystack_len:
            return -1

        for haystack_idx in range(haystack_len - needle_len + 1):
            isMatch: bool = True

            for needle_idx in range(needle_len):
                if haystack[haystack_idx + needle_idx] != needle[needle_idx]:
                    isMatch = False
                    break

            if isMatch:
                return haystack_idx

        return -1


    # Solution
    """
    # Solution 1
    #
    # Brute Force, Sliding Window
    #
    # Complexities:
    #   N - Length of `haystack`
    #   M - Length of `needle`
    #   - Time Complexity: O(N * M)
    #   - Space Complexity: O(1)
    """
    def solution1(self, haystack: str, needle: str) -> int:
        n = len(haystack)
        m = len(needle)

        if m > n:
            return -1

        for i in range(n - m + 1):
            if haystack[i: i + m] == needle:
                return i

        return -1

    """
    # Solution 2
    #
    # built-in: find()
    #
    # Complexities:
    #   N - Length of `haystack`
    #   M - Length of `needle`
    #   - Time Complexity: O(N * M)
    #   - Space Complexity: O(1)
    """
    def solution2(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)

File: leetcode/algorithms/28_find_the_index_of_the_first_occurrence_in_a_string/main.dart
class FindTheIndexOfTheFirstOccurrenceInAString {
  /**
   * Brute Force, Sliding Window
   *
   * Complexities:
   *   N - Length of `haystack`
   *   M - Length of `needle`
   *   - Time Complexity: O(N * M)
   *   - Space Complexity: O(1)
   */
  int strStr(String haystack, String needle) {
    int haystackLen = haystack.length;
    int needleLen = needle.length;

    if (needleLen > haystackLen) {
      return -1;
    }

    for (
      int haystackIdx = 0;
      haystackIdx <= haystackLen - needleLen;
      haystackIdx++
    ) {
      bool isMatch = true;

      for (int needleIdx = 0; needleIdx < needleLen; needleIdx++) {
        if (haystack[haystackIdx + needleIdx] != needle[needleIdx]) {
          isMatch = false;
          break;
        }
      }

      if (isMatch) {
        return haystackIdx;
      }
    }

    return -1;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Brute Force, Sliding Window
   *
   * Complexities:
   *   N - Length of `haystack`
   *   M - Length of `needle`
   *   - Time Complexity: O(N * M)
   *   - Space Complexity: O(1)
   */
  int solution1(String haystack, String needle) {
    int n = haystack.length;
    int m = needle.length;

    for (int i = 0; i <= n - m; i++) {
      int j = 0;

      while (j < m && haystack[i + j] == needle[j]) {
        j++;
      }

      if (j == m) {
        return i;
      }
    }

    return -1;
  }

  /**
   * Solution 2
   * 
   * built-in: indexOf()
   *
   * Complexities:
   *   N - Length of `haystack`
   *   M - Length of `needle`
   *   - Time Complexity: O(N * M)
   *   - Space Complexity: O(1)
   */
  int solution2(String haystack, String needle) {
    return haystack.indexOf(needle);
  }
}
